### DISCUSSION TITLE
Passing custom classes to actors

### DISCUSSION THREAD

**Discussion by NickCody** [actor-framework/actor-framework#1316] [Q&A] [answered]
I wrote this simple class to be passed to an actor, but:

```
static_assert((is_complete<type_id<strip_and_convert_t<Ts>>> && ...));
```
I thought I met the criteria set out in the docs:

- Inspectable (see Type Inspection)
- Default constructible
- Copy constructible

Here is the complete program. Any help would be greatly appreciated:

```
#include <iostream>

#include "caf/actor_system.hpp"
#include "caf/caf_main.hpp"
#include "caf/event_based_actor.hpp"

using namespace caf;

struct Foo {
  int bar;

  Foo() { bar = 0; }

  Foo(const Foo& foo) { bar = foo.bar }

  template <class Inspector> bool inspect(Inspector& f, Foo& x) { return f.object(x).fields(f.field("bar", x.bar)); };
};

// tried here too
//template <class Inspector> bool inspect(Inspector& f, Foo& x) { return f.object(x).fields(f.field("bar", x.bar)); };

behavior mirror(event_based_actor* self) {
  return {
    [=](const Foo& foo) { std::cout << foo.bar << std::endl; },
  };
}

void begin(event_based_actor* self, const actor& other) {
  Foo foo{ 1 };
  self->request(other, std::chrono::seconds(10), foo);
}

void caf_main(actor_system& sys) {
  auto mirror_actor = sys.spawn(mirror);
  sys.spawn(mirror);
}

CAF_MAIN()
```

**Comment by NickCody** [comment, marked as answer]
Ah, I see I can't use raw classes, they need to implement one of the actor based classes.

**Reply by Neverlord** [reply]
> Ah, I see I can't use raw classes, they need to implement one of the actor based classes.

I'm not sure what you mean by that. Subtyping actors is rarely needed. Maybe this article may help clearing up the "class-based vs function-based vs state-based" terminology: https://www.cafcademy.com/articles/implementing-actors-part-1.

> Here is the complete program. Any help would be greatly appreciated:

Two observations:
- `inspect` needs to be a free function, if you define it in the class body than it needs to be a `friend` function, not a member function.
- The compile-time check is asking whether there exists a type ID for `Foo`. For that you'll need a type ID block, like in this example here: https://github.com/actor-framework/actor-framework/blob/master/examples/custom_type/custom_types_1.cpp#L15.

Hope that helps. 🙂

**Comment by NickCody** [comment]
@Neverlord thanks so much for getting back to me. Yes, I got this working finally after reading the docs more thoroughly:

```
#include <iostream>

#include "caf/all.hpp"
#include "caf/actor_system.hpp"
#include "caf/caf_main.hpp"
#include "caf/event_based_actor.hpp"

using namespace caf;
using namespace std;

struct Foo;
struct Bar;

CAF_BEGIN_TYPE_ID_BLOCK(custom_types_1, first_custom_type_id)
CAF_ADD_TYPE_ID(custom_types_1, (Foo))
CAF_ADD_TYPE_ID(custom_types_1, (Bar))
CAF_END_TYPE_ID_BLOCK(custom_types_1)

struct Foo {
  int bar;
};

template <class Inspector> bool inspect(Inspector& f, Foo& x) {
  return f.object(x).fields(f.field("bar", x.bar));
}

struct Bar {
  int foo;
};

template <class Inspector> bool inspect(Inspector& f, Bar& x) {
  return f.object(x).fields(f.field("foo", x.foo));
}

behavior posh(event_based_actor* self) {
  return {
    [=](const Foo& val) { aout(self) << "Foo = " << deep_to_string(val) << endl; },
    [=](const Bar& val) { aout(self) << "Bar = " << deep_to_string(val) << endl; },
  };
}

void caf_main(actor_system& sys) {
  Foo f{ 10 };
  Bar b{ 20 };
  auto posh_actor = sys.spawn(posh);
  scoped_actor self{ sys };
  self->send(posh_actor, f);
  self->send(posh_actor, b);
}

CAF_MAIN(id_block::custom_types_1)
```

I'm coming from an Akka background and just needed to steep myself a bit more. Thanks again!
